File: 12/Ising_2D_v2.py
import numpy as np
import matplotlib.pyplot as plt
import time
import numba as nb

@nb.jit(nb.types.UniTuple(nb.int64,2)(nb.float64, nb.int64, nb.int64[:,:]))
def sweep(beta,L,s):
    
    probs=np.array([0.5,np.exp(-4*beta),np.exp(-8*beta)])
    
    dMtot=0
    dEtot=0

    for i in range(L):
        for j in range(L):
            
            dE=2*s[i,j]*(s[i-1,j]+s[(i+1)%L,j]+s[i,j-1]+s[i,(j+1)%L])

            if dE>=0 and probs[dE//4]<np.random.random(): continue
            
            s[i,j]=-s[i,j]
            dMtot+=2*s[i,j]
            dEtot+=dE

    return dMtot,dEtot


def runMC(L,NT,T,Navg):

    s=np.ones((L,L),dtype=int)
    M=L**2
    E=-2*L**2
    
    Msamp=np.zeros((NT,Navg))
    Esamp=np.zeros((NT,Navg))

    for i in range(NT):
        print(i)
        beta=1/T[i]
    
        for j in range(Navg):
            dM,dE=sweep(beta,L,s)
            M+=dM
            E+=dE
            
        for j in range(Navg):
            dM,dE=sweep(beta,L,s)
            M+=dM
            E+=dE
            Msamp[i,j]=M/L**2
            Esamp[i,j]=E/L**2

    Mavg=np.average(Msamp,axis=1)
    Eavg=np.average(Esamp,axis=1)
    
    M2avg=np.average(Msamp**2,axis=1)
    E2avg=np.average(Esamp**2,axis=1)

    return Mavg,Eavg,M2avg,E2avg


L=32
NT=100
T=np.linspace(0.04,4,NT)
Navg=50

t1=time.perf_counter()
Mavg,Eavg,M2avg,E2avg = runMC(L,NT,T,Navg)
t2=time.perf_counter()
print(t2-t1,"s")

suscept=(M2avg-Mavg**2)/T
heatcap=(E2avg-Eavg**2)/T**2


plt.subplot(2,2,1)
plt.plot(T,Mavg)
plt.subplot(2,2,2)
plt.plot(T,Eavg)
plt.subplot(2,2,3)
plt.plot(T,suscept)
plt.subplot(2,2,4)
plt.plot(T,heatcap)
plt.show()

File: 12/Ising_2D_v2.jl
function sweep(beta,L,s)

  probs=[0.5,exp(-4*beta),exp(-8*beta)]

  dMtot=0
  dEtot=0

  for i in 1:L
    im=i-1; if i==1 im=L end
    ip=i+1; if i==L ip=1 end
    for j in 1:L
      jm=j-1; if j==1 jm=L end
      jp=j+1; if j==L jp=1 end
   
      dE=2*s[i,j]*(s[im,j]+s[ip,j]+s[i,jm]+s[i,jp])
   
      if dE>=0 && probs[1+div(dE,4)]<rand()
        continue
      end
  
      s[i,j]=-s[i,j]
      dMtot+=2*s[i,j]
      dEtot+=dE
  
    end
  end
 
  return dMtot,dEtot
end


function runMC(L,NT,T,Navg)

  s=ones(Int,L,L)
  M=L^2
  E=-2*L^2

  Msamp=zeros(NT,Navg)
  Esamp=zeros(NT,Navg)

  for i in 1:NT
    println(i)
    beta=1/T[i]
  
    for j in 1:Navg
      dM,dE=sweep(beta,L,s)
      M+=dM
      E+=dE
    end
  
    for j in 1:Navg
      dM,dE=sweep(beta,L,s)
      M+=dM
      E+=dE
      Msamp[i,j]=M/L^2
      Esamp[i,j]=E/L^2
    end
  
  end

  Mavg=sum(Msamp,dims=2)/Navg
  Eavg=sum(Esamp,dims=2)/Navg
  
  M2avg=sum(Msamp.^2,dims=2)/Navg
  E2avg=sum(Esamp.^2,dims=2)/Navg

  return Mavg,Eavg,M2avg,E2avg
end


L=32
NT=100
T=LinRange(0.04,4.0,NT)
Navg=50

t1=time()
Mavg,Eavg,M2avg,E2avg = runMC(L,NT,T,Navg)
t2=time()
println(t2-t1,"s")

suscept=(M2avg-Mavg.^2)./T
heatcap=(E2avg-Eavg.^2)./T.^2


using Plots
gr()

p1=plot(T,Mavg,label="M")
p2=plot(T,Eavg,label="E")
p3=plot(T,suscept,label="χ")
p4=plot(T,heatcap,label="C")

display(plot(p1,p2,p3,p4,layout=(2,2)))
